### PR TITLE
[Reference Only/Draft] gazebo_ros_control ROS2 port

### PR DESCRIPTION
**Pull request by ddengster**
The code changes in this PR was intended to work with a 'coffeebot' simulation in gazebo. It has lots of hacks, and is only meant to be an incomplete guide to porting to ROS2. Used in conjunction with https://github.com/ros-controls/ros2_controllers/pull/36
 
Consider commits after the 18th Feb 2020, or e474da1.

Things of note will be put in reviews.

TLDR notes for porters:
- https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1083#discussion_r412745218
- https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1083#discussion_r412746244
- https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1083#discussion_r412756514
- https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1083#discussion_r412771786

Update: Also used with moveit2 training https://github.com/ros-planning/moveit2/pull/187

### CODE REVIEW

File: gazebo_ros_control/include/gazebo_ros_control/default_robot_hw_sim.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+                           double *const upper_limit, double *const effort_limit,
+                           double *const vel_limit);
+
+  unsigned int n_dof_;
+
+#if 0 //@todo
+  hardware_interface::JointStateInterface    js_interface_;
+  hardware_interface::EffortJointInterface   ej_interface_;
+  hardware_interface::PositionJointInterface pj_interface_;
+  hardware_interface::VelocityJointInterface vj_interface_;
```
Comment (ddengster) [inline gazebo_ros_control/include/gazebo_ros_control/default_robot_hw_sim.h:118]: my coffeebot simulation uses a function in EffortJointInterface to translate trajectories from joint_trajectory_controller to effort values. Hence the rest is disabled.

File: gazebo_ros_control/include/gazebo_ros_control/default_robot_hw_sim.h [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+#endif
+#if 0 //@todo
+  joint_limits_interface::EffortJointSaturationInterface   ej_sat_interface_;
+  joint_limits_interface::EffortJointSoftLimitsInterface   ej_limits_interface_;
+  joint_limits_interface::PositionJointSaturationInterface pj_sat_interface_;
+  joint_limits_interface::PositionJointSoftLimitsInterface pj_limits_interface_;
+  joint_limits_interface::VelocityJointSaturationInterface vj_sat_interface_;
+  joint_limits_interface::VelocityJointSoftLimitsInterface vj_limits_interface_;
```
Comment (ddengster) [inline gazebo_ros_control/include/gazebo_ros_control/default_robot_hw_sim.h:126]: Similar to the above, I only use a function from EffortJointSaturationInterface to clamp my effort values.

File: gazebo_ros_control/include/gazebo_ros_control/default_robot_hw_sim.h [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+#endif
+  std::vector<std::string> joint_names_;
+  std::vector<int> joint_types_;
+  std::vector<double> joint_lower_limits_;
+  std::vector<double> joint_upper_limits_;
+  std::vector<double> joint_effort_limits_;
+  std::vector<double> joint_vel_limits_;
+  std::vector<ControlMethod> joint_control_methods_;
+#if 0
+  std::vector<control_toolbox::Pid> pid_controllers_;
```
Comment (ddengster) [inline gazebo_ros_control/include/gazebo_ros_control/default_robot_hw_sim.h:136]: I havent found use for control_toolbox::Pid in gazebo_ros_controls, hence I disabled it. However, in another joint_trajectory_controller draft I used control_toolbox::Pid as part the code for converting trajectories to effort values. I'll link the joint_trajectory_controller PR later.

File: gazebo_ros_control/include/gazebo_ros_control/gazebo_ros_control_plugin.h [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  // Robot simulator interface
+  std::string robot_hw_sim_type_str_;
+  std::shared_ptr<gazebo_ros_control::RobotHWSim> robot_hw_sim_;
+  rclcpp::executors::MultiThreadedExecutor::SharedPtr executor_;
+  std::thread thread_executor_spin_; //executor_->spin causes lockups, us ethis alternative for now
+
+  // Controller manager
+  std::shared_ptr<controller_manager::ControllerManager> controller_manager_;
+  std::shared_ptr<controller_interface::ControllerInterface> controller_;
+  std::shared_ptr<controller_interface::ControllerInterface> controller2_;
```
Comment (ddengster) [inline gazebo_ros_control/include/gazebo_ros_control/gazebo_ros_control_plugin.h:121]: These 2 controllers are custom code. Normally you'd get the controller_manager to spawn controllers but this functionality is not available as of this writing. Might be wise to wait for that functionality before porting.

File: gazebo_ros_control/include/gazebo_ros_control/robot_hw_sim.h [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    JointData(const std::string& name, const std::string& hardware_interface) :
+      name_(name),
+      hardware_interface_(hardware_interface)
+    {}
+  };
+
+  /// \brief Gazebo plugin version of RobotHW
+  ///
+  /// An object of class RobotHWSim represents a robot's simulated hardware.
+  class RobotHWSim : public hardware_interface::RobotHardware
```
Comment (ddengster) [inline gazebo_ros_control/include/gazebo_ros_control/robot_hw_sim.h:68]: RobotHWSim used to inherit from hardware_interface::RobotHW, but in ros2 it has changed to hardware_interface::RobotHardware (and has different functionality to account for)

File: gazebo_ros_control/include/gazebo_ros_control/robot_hw_sim.h [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    /// \brief Set the emergency stop state
+    ///
+    /// Set the simulated robot's emergency stop state. The default implementation of this function does nothing.
+    ///
+    /// \param active  \c true if the emergency stop is active, \c false if not.
+    virtual void eStopActive(const bool active) {}
+
+    //@note: Avoid using these functions! ros2_control was not built with gazebo_ros_control 
+    //in mind, keep to using readSim and writeSim so we don't have to update the documents for now
+    virtual hardware_interface::hardware_interface_ret_t init() { }
```
Comment (ddengster) [inline gazebo_ros_control/include/gazebo_ros_control/robot_hw_sim.h:117]: as per comment, also check above.

File: gazebo_ros_control/src/default_robot_hw_sim.cpp [hunk 7]
```diff
@@ -0,0 +1,654 @@
...
+  ej_sat_interface_.enforceLimits(period);
+  ej_limits_interface_.enforceLimits(period);
+  pj_sat_interface_.enforceLimits(period);
+  pj_limits_interface_.enforceLimits(period);
+  vj_sat_interface_.enforceLimits(period);
+  vj_limits_interface_.enforceLimits(period);
+#else
+  for (auto eff_limit_handle : joint_eff_limit_handles_)
+  {
+    eff_limit_handle.enforceLimits(period);
```
Comment (ddengster) [inline gazebo_ros_control/src/default_robot_hw_sim.cpp:388]: limits handles applied here.

File: gazebo_ros_control/src/default_robot_hw_sim.cpp [hunk 8]
```diff
@@ -0,0 +1,654 @@
...
+      *joint_type = urdf_joint->type;
+      // Get limits from the URDF file.
+      if (joint_limits_interface::getJointLimits(urdf_joint->name, joint_limit_nh, limits))
+        has_limits = true;
+      if (joint_limits_interface::getSoftJointLimits(urdf_joint->name, joint_limit_nh, soft_limits))
+        has_soft_limits = true;
+      
+      //@note (ddeng): these joint limits arent input into the node, so fetch them from the urdf
+      limits.min_position = urdf_joint->limits->lower;
+      limits.max_position = urdf_joint->limits->upper;
```
Comment (ddengster) [inline gazebo_ros_control/src/default_robot_hw_sim.cpp:530]: This function DefaultRobotHWSim::registerJointLimits needs some trimming; it has a bunch of repeat logic.

File: gazebo_ros_control/src/gazebo_ros_control_plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+
+    //feed coffeebot_controller/coffeebot_controllers.yaml into the controllers
+    std::string param_file = 
+      ament_index_cpp::get_package_share_directory("coffeebot_controller") +
+      "/params/coffeebot_controllers.yaml";
+    RCLCPP_ERROR(rclcpp::get_logger("ros_control_plugin"),"param file: %s", param_file.c_str());
+ 
+    load_params_from_yaml(controller_->get_lifecycle_node(), param_file, "coffeebot_arm_effort_controller");
+    if (controller2_)
+      load_params_from_yaml(controller2_->get_lifecycle_node(), param_file, "coffeebot_gripper_effort_controller");
```
Comment (ddengster) [inline gazebo_ros_control/src/gazebo_ros_control_plugin.cpp:354]: Custom code: I load 2 controllers and associated parameter files here. Again, this functionality should be replaced by controller_manager spawn controller functionality when it has been implemented.

File: gazebo_ros_control/src/gazebo_ros_control_plugin.cpp [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+      }
+      else
+      {
+        reset_ctrlrs = false;
+      }
+    }
+
+    controller_->get_lifecycle_node()->set_parameter(rclcpp::Parameter("period", sim_period.seconds()));
+    if (controller2_)
+      controller2_->get_lifecycle_node()->set_parameter(rclcpp::Parameter("period", sim_period.seconds()));
```
Comment (ddengster) [inline gazebo_ros_control/src/gazebo_ros_control_plugin.cpp:439]: In ROS1, iirc each controller derives off Gazebo::Node, which supplies the "period" parameter automatically. In ROS2, each controller derives off LifecycleNode and the functionality interface with Gazebo::Node hasn't been done yet. Porters should take this into account.

File: gazebo_ros_control/src/default_robot_hw_sim.cpp [hunk 11]
```diff
@@ -0,0 +1,654 @@
...
+    limits.max_effort = joint_effort_limits_[j];
+    limits.has_effort_limits = true;
+
+    joint_pos_limit_handles_[j] = joint_limits_interface::PositionJointSaturationHandle(
+      joint_states_[j], joint_cmds_[j], limits);
+
+    joint_eff_cmdhandle_[j] = hardware_interface::JointCommandHandle(
+      joint_names_[j], &joint_effort_command_[j]);
+    //should be register_joint_effort_command_handle, but there's only 1 buffer for now
+    if (register_joint_command_handle(&joint_eff_cmdhandle_[j]) != hardware_interface::HW_RET_OK) {
```
Comment (ddengster) [inline gazebo_ros_control/src/default_robot_hw_sim.cpp:310]: The joint commands are registered here. The controllers (joint_trajectory_controller in this case) then use the registered commands. Porters should account for how these values are fed into the controllers.

File: gazebo_ros_control/include/gazebo_ros_control/default_robot_hw_sim.h (hunk 3 above)
Comment (ddengster) [reply gazebo_ros_control/include/gazebo_ros_control/default_robot_hw_sim.h:136]: https://github.com/ros-controls/ros2_controllers/pull/36

**Comment by chapulina** [comment]
Closing this in favor of https://github.com/ros-simulation/gazebo_ros2_control

**Pull request closed without merging**